### PR TITLE
[DOC] Add a notebook to describe inheritance structure of base classes

### PR DESCRIPTION
**Pull request by TonyBagnall**
adds a simplified overview of the core class structure of aeon, based on a simple UML by @hadifawaz1999 

I've presented UML with stylised/simplifed code. Happy to do otherwise if anyone has any ideas

I have as a result tweaked a couple of base classes, making fit/predict final and alterning some docstrings to standardise things

### CODE REVIEW

File: examples/base/img/uml.png
Comment (dguijo) [inline examples/base/img/uml.png:1]: There are three BaseDeepClusterer. The second and third should be Regressor and Classifier, respectively.

File: examples/base/img/uml3.png
Comment (dguijo) [inline examples/base/img/uml3.png:1]: There are three BaseDeepClusterer. The second and third should be Regressor and Classifier, respectively.

File: examples/base/base_classes.ipynb [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+   "source": [
+    "## sklearn `BaseEstimator` and aeon `BaseObject`\n",
+    "\n",
+    "To make sense of this, we break it down from the top.\n",
+    "\n",
+    "<img src=\"img/uml1.png\" width=\"600\" alt=\"Top level class hierarchy\">\n",
+    "\n",
+    "Everything inherits from sklearns `BaseEstimator`. This handles the mechanisms for\n",
+    "getting and setting parameters. The code structure below is stylised to show the\n",
+    "main functionsality and may differ in details from the actual implementations.\n",
```
Comment (dguijo) [inline examples/base/base_classes.ipynb:29]: functionsality -> functionality

File: examples/base/base_classes.ipynb [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+  {
+   "cell_type": "markdown",
+   "source": [
+    "## aeons ``BaseEstimator``, `BaseDeepNetwork` and `BaseMetric`\n",
+    "\n",
+    "Three classes extend `BaseObject`: ``BaseEstimator``, `BaseDeepNetwork` and\n",
+    "`BaseMetric`.\n",
+    "\n",
+    "`BaseDeepNetwork` is the base class for all the deep learning networks defined in the\n",
+    "`networks` class. It has a single abstract method `build_network`.\n",
```
Comment (dguijo) [inline examples/base/base_classes.ipynb:58]: change "class" by "module"?

File: examples/base/base_classes.ipynb [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+   "metadata": {
+    "collapsed": false
+   }
+  },
+  {
+   "cell_type": "markdown",
+   "source": [
+    "## `BaseForecaster` (aeon.forecasting.base)\n",
+    "contains the forecasting specific methods. More details are available in the API\n",
+    "'LINK]. `BaseForecaster` has the following concrete methods:\n",
```
Comment (dguijo) [inline examples/base/base_classes.ipynb:110]: replace "API-LINK" with the link.

File: examples/base/base_classes.ipynb [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+    "<img src=\"img/base_forecaster.png\" width=\"700\" alt=\"Base forecaster\">"
+   ],
+   "metadata": {
+    "collapsed": false
+   }
+  },
+  {
+   "cell_type": "markdown",
+   "source": [
+    "## BaseSeriesAnnotator (aeon.annotation.base)\n",
```
Comment (dguijo) [inline examples/base/base_classes.ipynb:121]: `BaseSeriesAnnotator` to keep using the same format :smile:

File: examples/base/base_classes.ipynb [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+    "<img src=\"img/base_transformer.png\" width=\"700\" alt=\"Base transformer\">\n"
+   ],
+   "metadata": {
+    "collapsed": false
+   }
+  },
+  {
+   "cell_type": "markdown",
+   "source": [
+    "## BaseCollectionEstimator (aeon.base)\n",
```
Comment (dguijo) [inline examples/base/base_classes.ipynb:150]: `BaseCollectionEstimator` to keep using the same format 😄

File: examples/base/base_classes.ipynb [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+    "<img src=\"img/base_classifier.png\" width=\"700\" alt=\"Top level class hierarchy\">"
+   ],
+   "metadata": {
+    "collapsed": false
+   }
+  },
+  {
+   "cell_type": "markdown",
+   "source": [
+    "## BaseRegressor (aeon.regression)\n",
```
Comment (dguijo) [inline examples/base/base_classes.ipynb:191]: `BaseRegressor` to keep using the same format 😄

File: examples/base/base_classes.ipynb [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+  {
+   "cell_type": "markdown",
+   "source": [
+    "The subclasses of `BaseCollectionEstimator` are as follows\n",
+    "\n",
+    "<img src=\"img/uml3.png\" width=\"600\" alt=\"Top level class hierarchy\">\n",
+    "\n",
+    "they have similar interfaces, but they are not identical\n",
+    "\n",
+    "## BaseClassifier (aeon.classification)\n",
```
Comment (dguijo) [inline examples/base/base_classes.ipynb:173]: `BaseClassifier` to keep using the same format 😄

File: examples/base/base_classes.ipynb [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+    "\n",
+    "This is the base class for all classifiers. It uses the standard `fit`, `predict` and\n",
+    " `predict_proba` structure from `sklearn`. `fit` and\n",
+    " `predict` call the abstract methods `_fit` and `_predict` which are implemented in\n",
+    " the subclass to define the classification algorithm. All of the common format checking\n",
+    " and conversion is donw using the following final methods defined in\n",
```
Comment (dguijo) [inline examples/base/base_classes.ipynb:179]: donw -> down

File: examples/base/base_classes.ipynb [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+    "<img src=\"img/base_regressor.png\" width=\"700\" alt=\"Top level class hierarchy\">"
+   ],
+   "metadata": {
+    "collapsed": false
+   }
+  },
+  {
+   "cell_type": "markdown",
+   "source": [
+    "## BaseClusterer (Aeon.clustering)\n",
```
Comment (dguijo) [inline examples/base/base_classes.ipynb:206]: `BaseClusterer` to keep using the same format 😄

Comment (dguijo) [reply examples/base/base_classes.ipynb:206]: also, "aeon" instead of "Aeon"

**Comment by TonyBagnall** [comment]
> Thank you Tony! Great work!
> 
> PS. Sorry for being picky 😅

picky is good! :)

File: examples/base/img/uml.png
Comment (TonyBagnall) [reply examples/base/img/uml.png:1]: removed for now, since we dont actually describe base deeps

File: examples/base/base_classes.ipynb (hunk 4 above)
Comment (dguijo) [reply examples/base/base_classes.ipynb:120]: I think this one has to be done?

**Review by dguijo — CHANGES_REQUESTED** [review]
Tony, there was one comment left. A minor minor thing. Sorry and thank you!

**Comment by TonyBagnall** [comment]
> Tony, there was one comment left. A minor minor thing. Sorry and thank you!

done

File: aeon/clustering/base.py [hunk 10]
```diff
@@ -22,11 +23,12 @@ class BaseClusterer(BaseCollectionEstimator, ABC):
     """
 
     def __init__(self, n_clusters: int = None):
-        self.fit_time_ = 0
         self.n_clusters = n_clusters
```
Comment (MatthewMiddlehurst) [inline aeon/clustering/base.py:26]: Should this really be a base class parameter? Seems restrictive.

**Review by MatthewMiddlehurst — COMMENTED** [review]
Two comments for the base class changes. Rest will be via notebook viewer, probably.

File: aeon/clustering/base.py (hunk 10 above)
Comment (TonyBagnall) [reply aeon/clustering/base.py:26]: that is a question for @chrisholder and beyond the scope of this PR, but I dont see why it should not be a base class parameter with a default?

Comment (TonyBagnall) [reply aeon/clustering/base.py:26]: seems fine to me, given it has a default value.

File: examples/base/base_classes.ipynb [hunk 11]
```diff
@@ -0,0 +1,266 @@
+{
```
[the remark this answers is not in the document]
Comment (TonyBagnall) [reply examples/base/base_classes.ipynb:1]: <p>not all estimators have fit and predict</p>

File: aeon/clustering/base.py (hunk 10 above)
Comment (TonyBagnall) [reply aeon/clustering/base.py:26]: but maybe raise an issue? Dont really want to hold design discussions on this PR which is just describing the current state

File: examples/base/base_classes.ipynb (hunk 11 above)
[the remark this answers is not in the document]
Comment (MatthewMiddlehurst) [reply examples/base/base_classes.ipynb:1]: The line it's referring to fit and predict as functions though I assumed. Won't push for it.

File: aeon/clustering/base.py (hunk 10 above)
Comment (MatthewMiddlehurst) [reply aeon/clustering/base.py:26]: ok

File: examples/base/base_classes.ipynb (hunk 11 above)
[the remark this answers is not in the document]
Comment (TonyBagnall) [reply examples/base/base_classes.ipynb:1]: ah I cant really see the output properly. Maybe you can tweak it later, might be easier

[the remark this answers is not in the document]
Comment (MatthewMiddlehurst) [reply examples/base/base_classes.ipynb:1]: https://app.reviewnb.com/aeon-toolkit/aeon/blob/ajb/base_notebook/examples/base/base_classes.ipynb/ seems to render fine

**Review by dguijo — APPROVED** [review]
LGTM, Tony and Matt. Thank you for your work!

**Pull request merged**